### PR TITLE
Make mbedtls windows ready, force CLANG on sgx, rearrange CI scripts

### PR DESCRIPTION
**Pull request by kwakkel1000**
We have fixes in CLANG for SGX, so now we force that you use CLANG if you compile for SGX.
We can now compile mbedtls on windows (with CI checked)
CI scripts are restructured for better understanding.

### CODE REVIEW

File: mbedtls-sys/Cargo.toml [hunk 1]
```diff
@@ -26,6 +26,9 @@ libc = { version = "0.2.0", optional = true }
 [target.'cfg(unix)'.dependencies]
 libc = { version = "0.2.0" }
 
+[target.'cfg(windows)'.dependencies]
+libc = { version = "0.2.0" }
```
Comment (Taowyoo) [inline mbedtls-sys/Cargo.toml:30]: Just a small suggestion, you could consider merging the following two cfg blocks into one:
```suggestion
[target.'cfg(any(unix, windows))'.dependencies]
libc = { version = "0.2.0" }
```

**Review by Taowyoo — APPROVED** [review]
LGTM

File: ci.sh [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+cd "$(dirname "$0")"
+
+repo_root=$(readlink -f $(dirname "${BASH_SOURCE[0]}"))
+
+if [ -z $TRAVIS_RUST_VERSION ]; then
+    echo "Expected TRAVIS_RUST_VERSION to be set in env"
+    exit 1
+fi
+
+python3 -m pip install -r ./mbedtls-sys/vendor/scripts/basic.requirements.txt
```
Comment (arai-fortanix) [inline ci.sh:12]: Can this use a python virtualenv instead of installing directly on the system?

File: mbedtls-platform-support/src/rust_printf.c [hunk 3]
```diff
@@ -22,7 +28,7 @@ extern int mbedtls_printf(const char *fmt, ...) {
        return -1;
 
     n++;
-    char p[n];
+    char *p = alloca(n);
```
Comment (arai-fortanix) [inline mbedtls-platform-support/src/rust_printf.c:31]: So this comment isn't specific to the change you're making here, since you're just changing a GNU variable-length array on the stack to an explicit `alloca()`, which is equivalent. But it's not a good idea to use either a variable-length array on the stack or an `alloca()` with an unbounded size. This allocation is on the stack, and neither the variable-length array allocation nor the `alloca()` can actually fail. This can result in stack overflow, which could result in security issues depending on the exact nature of the overflow.

There are basically two ways this could be fixed. Either setting an upper bound on the log message length (and simply truncating if the log message is longer than that), or allocating the buffer on the heap. If you're setting a maximum buffer size anad using the stack, then you may as well just use the fixed maximum size as the size of your array so you don't need `alloca()`. If you decide to heap allocate, it's probably a good idea to set a maximum size anyway, to reduce the possibility of heap exhaustion and/or attacks based on forcing a lot of heap to be allocated.

Basically: don't use `alloca()` or variable-sized arrays.

Comment (arai-fortanix) [reply mbedtls-platform-support/src/rust_printf.c:31]: Since this is already in the code, and your change doesn't make things any worse, I'm okay with fixing the stack usage in a follow-on change.

**Review by arai-fortanix — APPROVED** [review]

File: ci.sh (hunk 2 above)
Comment (Taowyoo) [reply ci.sh:15]: updated

File: mbedtls-platform-support/src/rust_printf.c (hunk 3 above)
Comment (Taowyoo) [reply mbedtls-platform-support/src/rust_printf.c:31]: Got, create an issue to keep tracking this : https://github.com/fortanix/rust-mbedtls/issues/317

**Comment by Taowyoo** [comment]
bors ping

**Comment by Taowyoo** [comment]
bors r=[arai-fortanix]
